src/transformers/modeling_utils.py

```diff
@@ -701,14 +702,38 @@ def load(module: nn.Module, state_dict, prefix=""):
                     with deepspeed.zero.GatheredParameters(params_to_gather, modifier_rank=0):
                         if torch.distributed.get_rank() == 0:
                             module._load_from_state_dict(*args)
-            else:
-                module._load_from_state_dict(*args)
 
         for name, child in module._modules.items():
             if child is not None:
                 load(child, state_dict, prefix + name + ".")
 
-    load(model_to_load, state_dict, prefix=start_prefix)
+    if is_deepspeed_zero3_enabled():
+        load(model_to_load, state_dict, prefix=start_prefix)
+    else:
+        # Adjust and remove our `start_prefix` as we don't need it anymore
+        state_dict = {
+            key[len(start_prefix) :] if key.startswith(start_prefix) else key: value
+            for key, value in state_dict.items()
+        }
+        # Finally we need to check if the params are the right dtype in the state dict
+        old_param = model_to_load
+        for param_name, param in state_dict.items():
+            splits = param_name.split(".")
+            old_param = model_to_load
+            for split in splits:
+                old_param = getattr(old_param, split)
+                if old_param is None:
+                    break
+            if old_param is not None:
+                if old_param.dtype != param.dtype:
+                    param = param.to(old_param.dtype)
+                if old_param.is_contiguous() and not param.is_contiguous():
+                    param = param.contiguous()
+                state_dict[param_name] = param
```

Comment:
Without this, we break the default behavior of always loading in `fp32`. 

Some example times with `llama-3-8B`:

Without specifying `torch_dtype`/loading in `fp32`:
Before: 2.95s
After: 2.85s

With specifying bfloat16 I still saw the 10x speedup :)

Reply:
what's the issue with fp32 (not specifying things)?

Reply:
Checking with Mark to see if lazy loading + upcast is *possible*

Reply:
that's interesting, so setting the default torch_dtype does an upcast? Or is it on the accelerate side? What I mean is loading bf16 into a f32 tensor with torch should be instand no?

Reply:
It is, the problem is doing so requires us loading the model weights into memory *right then*, and suddenly now we're no longer lazy!